File: src/bumperbot_mapping/bumperbot_mapping/mapping_with_known_poses.py
#!/usr/bin/env python3
import math
import rclpy
from rclpy.node import Node
from nav_msgs.msg import OccupancyGrid, MapMetaData
from sensor_msgs.msg import LaserScan
from tf2_ros import Buffer, TransformListener, LookupException
from tf_transformations import euler_from_quaternion

PROIOR_PROB = 0.5
OCC_PROB = 0.9
FREE_PROB = 0.35


class Pose:
    def __init__(self, px = 0, py = 0):
        self.x = px
        self.y = py

def coordinateToPose(px, py, map_info: MapMetaData):
    pose = Pose()
    pose.x = round((px - map_info.origin.position.x) / map_info.resolution)
    pose.y = round((py - map_info.origin.position.y) / map_info.resolution)
    return pose
    
def poseOnMap(pose: Pose, map_info: MapMetaData):
    return pose.x < map_info.width and pose.x >= 0 and pose.y < map_info.height and pose.y >= 0

def poseToCell(pose: Pose, map_info: MapMetaData):
    return map_info.width * pose.y + pose.x

def bresenham(start: Pose, end: Pose):
    line = []
    dx = end.x - start.x
    dy = end.y - start.y
    xsign = 1 if dx > 0 else -1
    ysign = 1 if dy > 0 else -1
    dx = abs(dx)
    dy = abs(dy)

    if dx > dy:
        xx = xsign
        xy = 0
        yx = 0
        yy = ysign
    else:
        tmp = dx
        dx = dy
        dy = tmp
        xx = 0
        xy = ysign
        yx = xsign
        yy = 0
    D = 2 * dy - dx
    y = 0

    for i in range(dx + 1):
        line.append(
            Pose(
                start.x + i * xx + y * yx,
                start.y + i * xy + y * yy
            )
        )
        if D >= 0:
            y += 1
            D -= 2 * dx
        D += 2 * dy
    return line

def inverseSensorMode(p_robot: Pose, p_beam: Pose): #type hint
    occ_values = []
    line = bresenham(p_robot, p_beam)
    occ_values.append[(line[-1]), 100]
    for pose in line[:-1]:
        occ_values.append((FREE_PROB))
    occ_values.append((line[-1], OCC_PROB))
    return occ_values

def prob2logodds(p):
    return math.log(p / (1 / p))
def logodd2prob(l):
    try:
        return 1 - (1 / (1 + math.exp(l)))
    except OverflowError:
        return 1.0 if l > 0 else 0.0


class MappingWithKnownPoses(Node):
    def __init__(self, name):
        super().__init__(name)

        self.declare_parameter("width", 50.0)
        self.declare_parameter("height", 50.0)
        self.declare_parameter("resolution",0.1)

        width = self.get_parameter("width").value
        height = self.get_parameter("height").value
        resolution = self.get_parameter("resolution").value

        self.map_ = OccupancyGrid()
        self.map_.info.resolution = resolution
        self.map_.info.width = round(width / resolution)
        self.map_.info.height = round(height / resolution)
        self.map_.info.origin.position.x = float(-round(width / 2.0))
        self.map_.info.origin.position.y = float(-round(height / 2.0))
        self.map_.header.frame_id = "odom"
        self.map_.data = [-1] *(self.map_.info.width * self.map_.info.height)

        self.probability_map_ = [prob2logodds(PROIOR_PROB)] * (self.map_.info.width * self.map_.info.height)


        self.map_pub = self.create_publisher(OccupancyGrid, "map", 1)
        self.scan_sub = self.create_subscription(LaserScan, "scan", self.scan_callback, 10)
        self.timer = self.create_timer(1.0, self.timer_callback)

        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

    def scan_callback(self, scan: LaserScan):
        try:
            t = self.tf_buffer.lookup_transform(self.map_.header.frame_id, scan.header.frame_id, rclpy.time.Time())
        except LookupException:
            self.get_logger().error("Unable to transform between /odom and /base_footprint")
            return
        robot_p = coordinateToPose(t.transform.translation.x, t.transform.translation.y, self.map_.info)
        if not poseOnMap(robot_p, self.map_.info):
            self.get_logger().error("The robot is out of the map!")
            return
        (roll, pitch, yaw) = euler_from_quaternion([t.transform.rotation.x, t.transform.rotation.y, t.transform.rotation.z, t.transform.rotation.w])
        
        for i in range(len(scan.ranges)): #distance between robot and obstacle follow by lazer[i]
            if math.isinf(scan.ranges[i]):
                continue
            #convert polar coordinates -> descartes coordinates
            angle = scan.angle_min + (i * scan.angle_increment) + yaw
            px = scan.ranges[i] * math.cos(angle)
            py = scan.ranges[i] * math.sin(angle)
            #/laser to /odom
            px += t.transform.translation.x
            py += t.transform.translation.y
            beam_p = coordinateToPose(px, py, self.map_.info)
            if not poseOnMap(beam_p, self.map_.info):
                continue
            
            poses = inverseSensorMode(robot_p, beam_p)
            for pose, value in poses:
                cell = poseToCell(pose, self.map_.info)
                self.probability_map_[cell] += prob2logodds(value) - prob2logodds(PROIOR_PROB)

    def timer_callback(self):
        self.map_.header.stamp = self.get_clock().now().to_msg()
        self.map_.data = [int(logodd2prob(value) * 100) for value in self.probability_map_]
        self.map_pub.publish(self.map_)


def main(args=None):
    rclpy.init(args=args)
    node = MappingWithKnownPoses("mapping_with_known_poses")
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: src/bumperbot_mapping/src/mapping_with_known_poses.cpp
#include "bumperbot_mapping/mapping_with_known_poses.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2/utils.h"

using namespace std::placeholders;

namespace bumperbot_mapping
{
//Function convert real coordinates(m) to grid coordinates
Pose coordinatesToPose(const double px, const double py, const nav_msgs::msg::MapMetaData & map_info)
{
    Pose pose;
    pose.x = std::round((px - map_info.origin.position.x) / map_info.resolution);
    pose.y = std::round((py - map_info.origin.position.y) / map_info.resolution);
    return pose;
}

//Funtion check location robot is in map ?
bool poseOnMap(const Pose & pose, const nav_msgs::msg::MapMetaData & map_info)
{
    return pose.x < static_cast<int>(map_info.width) && pose.x >= 0 &&
           pose.y < static_cast<int>(map_info.height) && pose.y >= 0;
}
//Convert 2D - > 1D
unsigned int poseToCell(const Pose &pose, const nav_msgs::msg::MapMetaData & map_info)
{
    return map_info.width * pose.y + pose.x;
}

std::vector<Pose> bresenham(const Pose &start, const Pose &end)
{
    // Implementation of Bresenham's line drawing algorithm
    // See en.wikipedia.org/wiki/Bresenham's_line_algorithm

    std::vector<Pose> line;

    int dx = end.x - start.x;
    int dy = end.y - start.y;

    int xsign = dx > 0 ? 1 : -1;
    int ysign = dy > 0 ? 1 : -1;

    dx = std::abs(dx);
    dy = std::abs(dy);
    int xx, xy, yx, yy;

    if (dx > dy)
    {
        xx = xsign;
        xy = 0;
        yx = 0;
        yy = ysign;
    }
    else
    {
        int tmp = dx;
        dx = dy;
        dy = tmp;

        xx = 0;
        xy = ysign;
        yx = xsign;
        yy = 0;
    }
    int D = 2 * dy - dx;
    int y = 0;

    line.reserve(dx + 1);
    for (int i = 0; i < dx + 1; i++)
    {
        line.emplace_back(
            Pose(
                start.x + i * xx + y * yx,
                start.y + i * xy + y * yy
            )
        );
        if (D >= 0)
        {
            y++;
            D -= 2 * dx;
        }

        D += 2 * dy;
    }
    return line;
}

std::vector<std::pair<Pose, double>>inverseSensorModel(const Pose & p_robot, const Pose & p_beam)
{
    std::vector<std::pair<Pose, double>> occ_values;
    std::vector<Pose> line = bresenham(p_robot, p_beam);
    occ_values.reserve(line.size());
    for (size_t i = 0; i < line.size() - 1u; i++)
    {
        occ_values.emplace_back(std::pair<Pose, double>(line.at(i), FREE_PROB));
    }
    occ_values.emplace_back(std::pair<Pose, double>(line.back(), OCC_PROB));
    return occ_values;
}

double prob2logodds(double p)
{
    return std::log((p) / (1 - p));
}

double logodds2prob(double l)
{
    return 1 - 1 / (1 + std::exp(l));
}

MappingWithKnownPoses::MappingWithKnownPoses(const std::string & name) : Node(name)
{
    declare_parameter<double>("width", 50.0);
    declare_parameter<double>("height", 50.0);
    declare_parameter<double>("resolution", 0.1);

    double width = get_parameter("width").as_double();
    double height = get_parameter("height").as_double();
    map_.info.resolution = get_parameter("resolution").as_double();
    map_.info.width = std::round(width / map_.info.resolution);
    map_.info.height = std::round(height / map_.info.resolution);
    map_.info.origin.position.x = - std::round(width / 2.0);
    map_.info.origin.position.y = - std::round(height / 2.0);
    map_.header.frame_id = "odom";
    map_.data = std::vector<int8_t>(map_.info.width * map_.info.height, -1);

    probability_map_ = std::vector<double>(map_.info.width * map_.info.height, prob2logodds(PRIOR_PROB));

    map_pub_ = create_publisher<nav_msgs::msg::OccupancyGrid>("map", 1);
    scan_sub_ = create_subscription<sensor_msgs::msg::LaserScan>("scan", 10, std::bind(&MappingWithKnownPoses::scanCallback, this, _1));
    timer_ = create_wall_timer(std::chrono::seconds(1), std::bind(&MappingWithKnownPoses::timerCallback, this));

    tf_buffer_ = std::make_unique<tf2_ros::Buffer>(get_clock());
    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
}
void MappingWithKnownPoses::scanCallback(const sensor_msgs::msg::LaserScan::SharedPtr scan)
{
    geometry_msgs::msg::TransformStamped t;
    try{
        t = tf_buffer_->lookupTransform(map_.header.frame_id, scan->header.frame_id, tf2::TimePointZero); //robot location in /odom frame
    }
    catch(tf2::TransformException &exc)
    {
        RCLCPP_ERROR(get_logger(), "Unable to transform between /odom and /base_footprint");
        return;
    }

    Pose robot_p = coordinatesToPose(t.transform.translation.x, t.transform.translation.y, map_.info); //(x, y) -> (cell_x, cell_y)
    if(!poseOnMap(robot_p, map_.info)) //check robot is in the map
    {
        RCLCPP_ERROR(get_logger(), "The robot is out of the Map");
        return;
    }

    tf2::Quaternion q(t.transform.rotation.x, t.transform.rotation.y, t.transform.rotation.z, t.transform.rotation.w);
    tf2::Matrix3x3 m(q);
    double roll, pitch, yaw;
    m.getRPY(roll, pitch, yaw);


    for(size_t i = 0; i< scan->ranges.size(); i++)
    {
        double angle = scan->angle_min + i * scan->angle_increment + yaw;
        double px = scan->ranges.at(i) * std::cos(angle);
        double py = scan->ranges.at(i) * std::sin(angle);
        px += t.transform.translation.x;
        py += t.transform.translation.y;
        Pose beam_p = coordinatesToPose(px, py, map_.info);
        if(!poseOnMap(beam_p, map_.info))
        {
            continue;
        }
        std::vector<std::pair<Pose, double>> poses = inverseSensorModel(robot_p, beam_p);
        for(const auto & pose : poses){
            if(poseOnMap(pose.first, map_.info)){
                unsigned int cell = poseToCell(pose.first, map_.info);
                probability_map_.at(cell) += prob2logodds(pose.second) - prob2logodds(PRIOR_PROB);
            }
        }
    }
}
//Publish map
void MappingWithKnownPoses::timerCallback()
{
    map_.header.stamp = get_clock()->now();
    std::transform(probability_map_.begin(), probability_map_.end(), map_.data.begin(), [](double value){
        return logodds2prob(value) * 100;
    });
    map_pub_->publish(map_);
}
} // namespace bumperbot_mapping


int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<bumperbot_mapping::MappingWithKnownPoses>("mapping_with_known_poses");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}